### PR TITLE
Add a --quality switch to warmup_stats.

### PR DESCRIPTION
**Pull request by ltratt**
PR #97 showed that we can get meaningful performance changes in table generation if we're prepared to tolerate slightly lower quality stats. In that PR, I made a rough estimate of how inaccurate they might be. However, I neglected to consider that floating point numbers have a very uneven spread: 50% of FP numbers reside between 0 and 1. In other words, the error is going to increase as someone's benchmark timings increase, which is surprising.

This commit introduces a --quality option with two settings: "low" (and fast) and "high" (and slow). The default is "high". Relative to #97, "low" also reduces the bootstrapping iterations, as well as using less accurate floating point summing. Roughly speaking, "high" is about 3x slower than "low", but one can observe small changes (at the fifth decimal place onwards) in data between low and high. Most of the time this is probably OK, but for paper quality benchmarking, the slightly longer time is worth it for more accurate stats, particularly as "high" quality tends to be more stable across runs (whereas there can be minor differences between low quality runs, which can confuse users).

I'm attaching two tables (as a zip, because GitHub doesn't allow us to attach HTML... )built with high and low quality stats (based on the 1.5 run of the warmup experiment): the biggest change I can see is TruffleRuby/binarytree's mean changing from 2.61061 (high quality) to 2.61059 (low quality). [tables.zip](https://github.com/softdevteam/warmup_stats/files/3029480/tables.zip)

CC/@smarr

### CODE REVIEW

**Comment by smarr** [comment]
I see. I was thinking of making the number of bootstrap iterations configurable, too. Since it seems to be one of the main things dominating performance.

The next step would have been a similar random number generator with uniform distribution, but who knows what statisticians would think of that ;)

File: warmup/bootstrapper.py [hunk 1]
```diff
@@ -70,43 +70,68 @@ def _mean(data):
...
-            # so if you ever crank BOOTSTRAP_ITERATIONS to a mammoth value, keep this in mind.
+            # Note that summing into a float like this does cause rounding
+            # errors, but for 10,000 bootstrap iterations and values in the
+            # 0.0-1.0 range, that error is only apparent at the 10th decimal
+            # point. However, the error increases as the bootstrap iterations
+            # and/or timings increase above: roughly speaking the error becomes
+            # 1 decimal place worse for each order of magnitude bigger the
+            # bootstrap iterations and/or timings become.
             sample_sum = 0.0
-
```
Comment (vext01) [inline warmup/bootstrapper.py]: Total nitpick, but arguably that blank line change should be in a separate commit. Personally, I'd just have left it unless I was changing an adjacent line.

**Comment by ltratt** [comment]
In an ideal world, we'd convert the whole shebang to use PyPy, which would help a lot. Unfortunately, after I tried doing just that, I found out we'd need to convert the code to Python 3, since rpy2 will never work in Python 2.x mode on PyPy :/ This PR is a halfway house in that regard.

>The next step would have been a similar random number generator with uniform distribution, but who knows what statisticians would think of that ;)

Ha ;)

File: warmup/bootstrapper.py [hunk 2]
```diff
@@ -70,43 +70,68 @@ def _mean(data):
...
 
-            # Note that summing into a float like this does cause rounding errors, but for 100,000
-            # bootstrap iterations that error is only apparent at the 9th decimal point, which is
-            # acceptable in our case. Roughly speaking, each factor of 10 bigger that
-            # BOOTSTRAP_ITERATIONS becomes, the error becomes 1 decimal place "worse",
-            # so if you ever crank BOOTSTRAP_ITERATIONS to a mammoth value, keep this in mind.
+            # Note that summing into a float like this does cause rounding
+            # errors, but for 10,000 bootstrap iterations and values in the
+            # 0.0-1.0 range, that error is only apparent at the 10th decimal
+            # point. However, the error increases as the bootstrap iterations
```
Comment (vext01) [inline warmup/bootstrapper.py]: 10th decimal *place*? (I know what you mean of course).

File: warmup/bootstrapper.py [hunk 3]
```diff
@@ -70,43 +70,68 @@ def _mean(data):
...
                 num_samples += seg_len
                 for _ in xrange(seg_len):
                     sample_sum += seg[int(random.random() * seg_len)]
 
             means.append(sample_sum / float(num_samples))
-    assert len(means) >= BOOTSTRAP_ITERATIONS
+    assert len(means) >= BOOTSTRAP_ITERATIONS_LOWQ
+    return means
+
+def _bootstrap_means_highq(steady_segments_all_pexecs):
```
Comment (vext01) [inline warmup/bootstrapper.py]: Did you put the high quality code in a separate function for performance purposes? To avoid repeated "am I hi quality" conditionals?

File: warmup/bootstrapper.py [hunk 4]
```diff
@@ -129,9 +154,12 @@ def bootstrap_steady_perf(steady_segments_all_pexecs, confidence_level=CONFIDENC
 
 
 if __name__ == '__main__':
-    import json
-    import sys
+    parser = argparse.ArgumentParser(description='Bootstrap data.')
```
Comment (vext01) [inline warmup/bootstrapper.py]: Is there any reason the argument is upper-case?

File: warmup/bootstrapper.py (hunk 1 above)
Comment (ltratt) [reply warmup/bootstrapper.py]: It's to make the two functions in the file more easily comparable. I don't think it's very important either way, though.

File: warmup/bootstrapper.py (hunk 2 above)
Comment (ltratt) [reply warmup/bootstrapper.py]: Oops!

File: warmup/bootstrapper.py (hunk 3 above)
Comment (ltratt) [reply warmup/bootstrapper.py]: The latter.

File: warmup/bootstrapper.py (hunk 4 above)
Comment (ltratt) [reply warmup/bootstrapper.py]: Not sure I understand.

**Comment by smarr** [comment]
Doing everything in PyPy isn't much of a win. I did the PyPy conversion for everything but the RPy2 stuff. Didn't see much benefit.
Bootstrapping is what actually shows up in a profile.
Also, the plots require matplotlib, which is currently not PyPy compatible, when I tried.
There PyPy might help, didn't profile that.

File: warmup/bootstrapper.py (hunk 1 above)
Comment (vext01) [reply warmup/bootstrapper.py]: As long as it was a concious decision I don't mind either way. I was expecting one function with conditionals, and perhaps assumed you were doing something cleverer than you were :)

File: warmup/bootstrapper.py (hunk 3 above)
Comment (vext01) [reply warmup/bootstrapper.py]: As long as it was a concious decision I don't mind either way. I was expecting one function with conditionals, and perhaps assumed you were doing something cleverer than you were :)

(I moved this comment from above, as I originally posted in the wrong place, sorry)

File: warmup/bootstrapper.py (hunk 4 above)
Comment (vext01) [reply warmup/bootstrapper.py]: I meant to comment on the following line, where we have 'HIGH' in all caps. Why not 'high'? Maybe you want the user to really understand the implications of deviating from the default?

**Comment by ltratt** [comment]
matplotlib has worked in PyPy for a while (though it's a mild pain to install): and there are two versions of numpy, one of which is traceable (so *might* help in some parts). So fully converting to PyPy *might* help; but I agree that bootstrapping is probably the single biggest cost.

File: warmup/bootstrapper.py (hunk 3 above)
Comment (ltratt) [reply warmup/bootstrapper.py]: Note that the two functions have fairly different bodies: I think that making various bits conditional would make it hard to read.

File: warmup/bootstrapper.py (hunk 4 above)
Comment (ltratt) [reply warmup/bootstrapper.py]: I'm just fitting in with the surrounding code (e.g. "debug"). I agree that it looks weird...

**Comment by vext01** [comment]
If Stefan is happy, then I think it's time to squash.

**Comment by ltratt** [comment]
Squashed.

**Comment by smarr** [comment]
I didn't review, but I support the idea of having a fast-run option. [so, this may satisfy the "happy" requirement from my side]
